File: commonutils/src/main/java/net/maxsmr/commonutils/android/AppUtils.java
package net.maxsmr.commonutils.android;

import android.app.ActivityManager;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.net.Uri;
import android.os.Build;
import android.os.StrictMode;
import android.support.v4.content.FileProvider;
import android.text.TextUtils;
import android.webkit.MimeTypeMap;

import net.maxsmr.commonutils.android.processmanager.AbstractProcessManager;
import net.maxsmr.commonutils.android.processmanager.model.ProcessInfo;
import net.maxsmr.commonutils.data.CompareUtils;
import net.maxsmr.commonutils.data.FileHelper;
import net.maxsmr.commonutils.data.ReflectionUtils;
import net.maxsmr.commonutils.logger.BaseLogger;
import net.maxsmr.commonutils.logger.holder.BaseLoggerHolder;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.lang.reflect.Method;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

public final class AppUtils {

    private static final BaseLogger logger = BaseLoggerHolder.getInstance().getLogger(AppUtils.class);

    private AppUtils() {
        throw new AssertionError("no instances");
    }

    public static boolean isPackageInstalled(@NotNull Context context, String packageName) {
        return getApplicationInfo(context, packageName) != null;
    }

    @Nullable
    public static ApplicationInfo getApplicationInfo(@NotNull Context context, String packageName) {
        return getApplicationInfo(context, packageName, 0);
    }

    @Nullable
    public static ApplicationInfo getApplicationInfo(@NotNull Context context, String packageName, int flags) {
        PackageManager packageManager = context.getPackageManager();
        try {
            return packageManager.getApplicationInfo(packageName, flags);
        } catch (PackageManager.NameNotFoundException e) {
            return null;
        }
    }

    @Nullable
    public static PackageInfo getPackageInfo(@NotNull Context context, String packageName) {
        return getPackageInfo(context, packageName, 0);
    }

    @Nullable
    public static PackageInfo getPackageInfo(@NotNull Context context, String packageName, int flags) {
        PackageManager packageManager = context.getPackageManager();
        if (!TextUtils.isEmpty(packageName)) {
            try {
                return packageManager.getPackageInfo(packageName, flags);
            } catch (PackageManager.NameNotFoundException e) {
            }
        }
        return null;
    }

    @Nullable
    public static PackageInfo getArchivePackageInfo(@NotNull Context context, File apkFile) {
        return getArchivePackageInfo(context, apkFile, 0);
    }

    @Nullable
    public static PackageInfo getArchivePackageInfo(@NotNull Context context, File apkFile, int flags) {
        if (FileHelper.isFileCorrect(apkFile)) {
            PackageManager packageManager = context.getPackageManager();
            return packageManager.getPackageArchiveInfo(apkFile.getAbsolutePath(), flags);
        }
        return null;
    }

    @Nullable
    public static Intent getLaunchIntentForPackage(@NotNull Context context, String packageName) {
        PackageManager packageManager = context.getPackageManager();
        if (!TextUtils.isEmpty(packageName)) {
            try {
                return packageManager.getLaunchIntentForPackage(packageName);
            } catch (Exception e) {
                logger.e("an Exception occurred during getLaunchIntentForPackage(): " + e.getMessage(), e);
            }
        }
        return null;
    }

    /**
     * @return {@linkplain Intent} only if given apk file is installed
     */
    @Nullable
    public static Intent getArchiveLaunchIntentForPackage(@NotNull Context context, File apkFile) {
        PackageInfo packageInfo = getArchivePackageInfo(context, apkFile);
        return packageInfo != null ? getLaunchIntentForPackage(context, packageInfo.packageName) : null;
    }

    @Nullable
    public static Intent getSelfLaunchIntentForPackage(@NotNull Context context) {
        return getLaunchIntentForPackage(context, context.getPackageName());
    }

    public static void launchSelf(@NotNull Context context) {
        Intent activityIntent = AppUtils.getSelfLaunchIntentForPackage(context);
        if (activityIntent == null) {
            throw new RuntimeException("Cannot get launch intent for " + context.getPackageName());
        }
        activityIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        context.startActivity(activityIntent);
    }

    /**
     * @param fileUriProviderAuthority specify string after packageName + ".", if intended to use FileProvider instead of file://
     * */
    @Nullable
    public static Intent getViewFileIntent(@NotNull Context context, @Nullable File file, @Nullable String fileUriProviderAuthority) {
        Intent result = null;
        if (FileHelper.isFileExists(file)) {
            result = new Intent(Intent.ACTION_VIEW);

            final Uri fileUri;

            if (!TextUtils.isEmpty(fileUriProviderAuthority)) {
                fileUri= FileProvider.getUriForFile(context, context.getPackageName() + "." + fileUriProviderAuthority, file);
            } else {
                fileUri = Uri.fromFile(file);
            }

            final String mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(FileHelper.getFileExtension(file));
            result.setDataAndType(fileUri, mimeType);

            result.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);

            if (!TextUtils.isEmpty(fileUriProviderAuthority)) {
                result.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
            }
            if (!canHandleActivityIntent(context, result)) {
                result = null;
            }
        }
        return result;
    }

    /** try to disable throwing exception when sharing file:// scheme instead of FileProvider */
    public static boolean disableFileUriStrictMode() {
        boolean result = true;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            result = false;
            Method disableMethod = null;
            try {
                disableMethod = ReflectionUtils.findMethod(StrictMode.class, "disableDeathOnFileUriExposure");
            } catch (NoSuchMethodException e) {
                logger.e("Cannot disable \"death on file uri exposure\": " + e.getMessage(), e);
            }
            if (disableMethod != null) {
                try {
                    disableMethod.invoke(null);
                    result = true;
                } catch (Exception e) {
                    logger.e("Cannot disable \"death on file uri exposure\": " + e.getMessage(), e);
                }
            }
        }
        return result;
    }

    public static boolean canHandleActivityIntent(@NotNull Context context, @Nullable Intent intent) {
        List<ResolveInfo> resolveInfos = intent != null ? context.getPackageManager().queryIntentActivities(intent, 0) : null;
        return resolveInfos != null && !resolveInfos.isEmpty();
    }

    public static int getPackageActivitiesCount(@NotNull PackageInfo packageInfo) {
        if (packageInfo.activities == null || packageInfo.activities.length == 0) {
            return 0;
        }
        return packageInfo.activities.length;
    }


    /**
     * Get APK build code
     *
     * @param context Context to use
     * @return Build version or 0 if app is not installed
     */
    public static int getApkVersionCode(@NotNull Context context, File apkFile) {
        PackageInfo packageInfo = getArchivePackageInfo(context, apkFile);
        if (packageInfo != null) {
            return packageInfo.versionCode;
        }
        return 0;
    }

    /**
     * Get installed package build code
     *
     * @param context Context to use
     * @return Build version or 0 if app is not installed
     */
    public static int getAppVersionCode(@NotNull Context context, String packageName) {
        PackageInfo packageInfo = getPackageInfo(context, packageName);

        if (packageInfo != null) {
            return packageInfo.versionCode;
        }
        return 0;
    }

    /**
     * Check if app requires update
     */
    public static boolean requiresUpdate(@NotNull Context context, String packageName, int build) {
        int currentVersionCode = getAppVersionCode(context, packageName);
        logger.i("Package: " + packageName + ", current: " + currentVersionCode + ", new: " + build);
        return currentVersionCode > 0 && build > 0 && currentVersionCode != build || (currentVersionCode <= 0 || build <= 0);
    }


    /**
     * Get application title by package name
     *
     * @return null if not found
     */
    @Nullable
    public static String getAppTitle(@NotNull Context context, String packageName) {
        PackageManager packageManager = context.getPackageManager();
        ApplicationInfo info = getApplicationInfo(context, packageName);
        return info != null ? info.loadLabel(packageManager).toString() : null;
    }

    /**
     * @return uid for specified name or -1 if not found
     */
    public static int getApplicationUid(@NotNull Context context, String packageName) {
        ApplicationInfo info = getApplicationInfo(context, packageName);
        return info != null ? info.uid : -1;
    }

    public static boolean isSystemApp(@Nullable PackageInfo packageInfo) {
        return isSystemApp(packageInfo != null ? packageInfo.applicationInfo : null);
    }

    public static boolean isSystemApp(@Nullable ApplicationInfo applicationInfo) {
        return applicationInfo != null &&
                (applicationInfo.flags &
                        (ApplicationInfo.FLAG_UPDATED_SYSTEM_APP | ApplicationInfo.FLAG_SYSTEM)) > 0;
    }

    /**
     * @return null if not found
     */
    public static Boolean isAppInBackground(@NotNull Context context, String packageName) {
        ActivityManager am = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
        if (am == null) {
            throw new RuntimeException(ActivityManager.class.getSimpleName() + " is null");
        }
        Boolean isInBackground = null;
        if (!TextUtils.isEmpty(packageName)) {
            List<ActivityManager.RunningAppProcessInfo> runningProcesses = am.getRunningAppProcesses();
            for (ActivityManager.RunningAppProcessInfo processInfo : runningProcesses) {
                if (packageName.equals(processInfo.processName)) {
                    isInBackground = processInfo.importance != ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND;
                }
            }
        }
        return isInBackground;
    }

    /**
     * @return null if not found
     */
    public static Boolean isAppInBackground(String packageName,
                                            @NotNull AbstractProcessManager manager, boolean includeSystemPackages) {
        Boolean isInBackground = null;
        if (!TextUtils.isEmpty(packageName)) {
            for (ProcessInfo info : manager.getProcesses(includeSystemPackages)) {
                if (CompareUtils.stringsEqual(info.packageName, packageName, false)) {
                    isInBackground = info.isForeground != null && !info.isForeground;
                    break;
                }
            }
        }
        return isInBackground;
    }


    @NotNull
    public static Set<Integer> getPidsByName(@Nullable String packageName,
                                             @NotNull AbstractProcessManager manager, boolean includeSystemPackages) {
        return getPidsByName(packageName, manager, includeSystemPackages, CompareUtils.MatchStringOption.EQUALS.flag);
    }

    @NotNull
    public static Set<Integer> getPidsByName(@Nullable String packageName,
                                             @NotNull AbstractProcessManager manager, boolean includeSystemPackages, int matchFlags) {
        Set<Integer> pids = new LinkedHashSet<>();
        if (!TextUtils.isEmpty(packageName)) {
            final List<ProcessInfo> runningProcesses = manager.getProcesses(includeSystemPackages);
            for (ProcessInfo process : runningProcesses) {
                if (CompareUtils.stringsMatch(packageName, process.packageName, matchFlags)) {
                    pids.add(process.pid);
                }
            }
        }
        return pids;
    }

}

File: commonutils/src/main/kotlin/net/maxsmr/commonutils/android/AppUtils.kt
package net.maxsmr.commonutils.android

import android.app.ActivityManager
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.content.pm.ApplicationInfo
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.os.Build
import android.os.StrictMode
import android.util.DisplayMetrics
import android.util.TypedValue
import android.view.WindowManager
import net.maxsmr.commonutils.android.processmanager.AbstractProcessManager
import net.maxsmr.commonutils.data.*
import net.maxsmr.commonutils.data.text.EMPTY_STRING
import net.maxsmr.commonutils.data.text.isEmpty
import net.maxsmr.commonutils.logger.BaseLogger
import net.maxsmr.commonutils.logger.holder.BaseLoggerHolder
import java.io.File
import java.lang.reflect.Method
import java.util.*

private val logger = BaseLoggerHolder.getInstance().getLogger<BaseLogger>("AppUtils")

fun isPackageInstalled(context: Context, packageName: String): Boolean {
    return getApplicationInfo(context, packageName) != null
}

fun getSelfVersionName(context: Context): String {
    return getVersionName(getSelfPackageInfo(context, PackageManager.GET_META_DATA))
}

fun getSelfVersionCode(context: Context): Long? {
    return getVersionCode(getSelfPackageInfo(context, PackageManager.GET_META_DATA))
}

fun getVersionName(context: Context, packageName: String): String {
    return getVersionName(getPackageInfo(context, packageName, PackageManager.GET_META_DATA))
}

fun getVersionCode(context: Context, packageName: String): Long? {
    return getVersionCode(getPackageInfo(context, packageName, PackageManager.GET_META_DATA))
}

fun getArchiveVersionName(context: Context, apkFile: File?): String {
    return getVersionName(getArchivePackageInfo(context, apkFile, PackageManager.GET_META_DATA))
}

fun getArchiveVersionCode(context: Context, apkFile: File?): Long? {
    return getVersionCode(getArchivePackageInfo(context, apkFile, PackageManager.GET_META_DATA))
}

fun getSelfApplicationInfo(context: Context, flags: Int = 0): ApplicationInfo? {
    return getApplicationInfo(context, context.packageName, flags)
}

fun getApplicationInfo(
        context: Context,
        packageName: String,
        flags: Int = 0
): ApplicationInfo? {
    val packageManager = context.packageManager
    return try {
        packageManager.getApplicationInfo(packageName, flags)
    } catch (e: PackageManager.NameNotFoundException) {
        null
    }

}

fun getSelfPackageInfo(context: Context, flags: Int = 0): PackageInfo? {
    return getPackageInfo(context, context.packageName)
}

fun getPackageInfo(
        context: Context,
        packageName: String,
        flags: Int = 0
): PackageInfo? {
    val packageManager = context.packageManager
    if (!isEmpty(packageName)) {
        try {
            return packageManager.getPackageInfo(packageName, flags)
        } catch (e: PackageManager.NameNotFoundException) {
            // ignored
        }

    }
    return null
}

fun getArchivePackageInfo(
        context: Context,
        apkFile: File?,
        flags: Int = 0
): PackageInfo? {
    if (apkFile != null && FileHelper.isFileCorrect(apkFile)) {
        val packageManager = context.packageManager
        return packageManager.getPackageArchiveInfo(apkFile.absolutePath, flags)
    }
    return null
}

fun getLaunchIntentForPackage(context: Context, packageName: String): Intent? {
    val packageManager = context.packageManager
    if (!isEmpty(packageName)) {
        try {
            return packageManager.getLaunchIntentForPackage(packageName)
        } catch (e: Exception) {
            logger.e("an Exception occurred during getLaunchIntentForPackage(): " + e.message, e)
        }
    }
    return null
}

/**
 * @return [Intent] only if given apk file is installed
 */
fun getArchiveLaunchIntentForPackage(context: Context, apkFile: File): Intent? {
    val packageInfo = getArchivePackageInfo(context, apkFile)
    return if (packageInfo != null) getLaunchIntentForPackage(context, packageInfo.packageName) else null
}

fun getSelfLaunchIntentForPackage(context: Context): Intent? {
    return getLaunchIntentForPackage(context, context.packageName)
}

fun launchSelf(context: Context, flags: Int = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK) {
    val activityIntent = getSelfLaunchIntentForPackage(context)
            ?: throw RuntimeException("Cannot launch: self package ${context.packageName} intent not found")
    if (flags != 0) {
        activityIntent.flags = flags
    }
    context.startActivity(activityIntent)
}

/**
 * try to disable throwing exception when sharing file:// scheme instead of [androidx.core.content.FileProvider]
 */
fun disableFileUriStrictMode(): Boolean {
    var result = true
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
        result = false
        var disableMethod: Method? = null
        try {
            disableMethod = ReflectionUtils.findMethod(StrictMode::class.java, "disableDeathOnFileUriExposure", false, true)
        } catch (e: NoSuchMethodException) {
            logger.e("Cannot disable \"death on file uri exposure\": " + e.message, e)
        }

        if (disableMethod != null) {
            try {
                disableMethod.invoke(null)
                result = true
            } catch (e: Exception) {
                logger.e("Cannot disable \"death on file uri exposure\": " + e.message, e)
            }

        }
    }
    return result
}

fun canLaunchIntent(context: Context, intent: Intent?): Boolean {
    var result = false
    if (intent != null) {
        val pm = context.packageManager
        if (pm.resolveActivity(intent, 0) != null) {
            result = true
        }
    }
    return result
}

fun canHandleActivityIntent(context: Context, intent: Intent?): Boolean {
    val resolveInfos = if (intent != null) context.packageManager.queryIntentActivities(intent, 0) else null
    return resolveInfos != null && resolveInfos.isNotEmpty()
}

fun getPackageActivitiesCount(packageInfo: PackageInfo): Int {
    return if (packageInfo.activities == null || packageInfo.activities.isEmpty()) {
        0
    } else {
        packageInfo.activities.size
    }
}

/**
 * Check if app requires update
 */
fun isPackageRequireUpdate(
        context: Context,
        packageName: String,
        targetVersionCode: Long
): Boolean {
    val currentVersionCode = getVersionCode(context, packageName)
    if (currentVersionCode == null) {
        logger.e("Version code for $packageName not found")
        return true
    }
    logger.i("Package: $packageName, current: $currentVersionCode, new: $targetVersionCode")
    return currentVersionCode > 0 && targetVersionCode > 0 && currentVersionCode != targetVersionCode
            || currentVersionCode <= 0 || targetVersionCode <= 0
}

fun getSelfApplicationLabel(context: Context): String? {
    return getApplicationLabel(context, context.packageName)
}

/**
 * Get application title by package name
 *
 * @return null if not found
 */
fun getApplicationLabel(context: Context, packageName: String): String? {
    val packageManager = context.packageManager
    val info = getApplicationInfo(context, packageName)
    return info?.loadLabel(packageManager)?.toString()
}

fun getApplicationUid(context: Context): Int? {
    return getApplicationUid(context, context.packageName)
}

/**
 * @return uid for specified name or null if not found
 */
fun getApplicationUid(context: Context, packageName: String): Int? {
    val info = getApplicationInfo(context, packageName)
    return info?.uid
}

fun isSelfAppInBackground(context: Context): Boolean {
    val result = isAppInBackground(context, context.packageName)
    return result
            ?: throw RuntimeException("Self package ${context.packageName} not found")
}

/**
 * @return null if not found
 */
fun isAppInBackground(context: Context, packageName: String): Boolean? {
    val am = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager?
            ?: throw RuntimeException(ActivityManager::class.java.simpleName + " is null")
    var isInBackground: Boolean? = null
    if (!isEmpty(packageName)) {
        val runningProcesses = am.runningAppProcesses
        if (runningProcesses != null) {
            for (processInfo in runningProcesses) {
                if (packageName == processInfo.processName) {
                    isInBackground = processInfo.importance != ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND
                }
            }
        }
    }
    return isInBackground
}

fun isSelfAppInBackground(
        context: Context,
        manager: AbstractProcessManager,
        includeSystemPackages: Boolean
): Boolean? {
    return isAppInBackground(context.packageName, manager, includeSystemPackages)
}

/**
 * @return null if not found
 */
fun isAppInBackground(
        packageName: String,
        manager: AbstractProcessManager,
        includeSystemPackages: Boolean
): Boolean? {
    var isInBackground: Boolean? = null
    if (!isEmpty(packageName)) {
        for (info in manager.getProcesses(includeSystemPackages)) {
            if (stringsEqual(info.packageName, packageName, false)) {
                isInBackground = info.isForeground != null && !info.isForeground
                break
            }
        }
    }
    return isInBackground
}

fun getPidsByName(
        packageName: String?,
        manager: AbstractProcessManager,
        includeSystemPackages: Boolean,
        matchFlags: Int = MatchStringOption.EQUALS.flag
): Set<Int> {
    val pids = LinkedHashSet<Int>()
    if (!isEmpty(packageName)) {
        val runningProcesses = manager.getProcesses(includeSystemPackages)
        for (process in runningProcesses) {
            if (stringsMatch(packageName, process.packageName, matchFlags)) {
                pids.add(process.pid)
            }
        }
    }
    return pids
}

fun getVersionName(info: PackageInfo?): String {
    return if (info != null) {
        info.versionName
    } else {
        EMPTY_STRING
    }
}

fun getVersionCode(info: PackageInfo?): Long? {
    return if (info != null) {
        if (SdkUtils.isAtLeastPie()) {
            info.longVersionCode
        } else {
            @Suppress("DEPRECATION")
            info.versionCode.toLong()
        }
    } else {
        null
    }
}

fun isSystemApp(packageInfo: PackageInfo?): Boolean? {
    return isSystemApp(packageInfo?.applicationInfo)
}

/**
 * @return null if not found
 */
fun isSystemApp(applicationInfo: ApplicationInfo?): Boolean? {
    if (applicationInfo == null) {
        return false;
    }
    return applicationInfo.flags and (ApplicationInfo.FLAG_UPDATED_SYSTEM_APP or ApplicationInfo.FLAG_SYSTEM) > 0
}

/**
 * Задает локаль по умолчанию в рамках приложения.
 *
 * @param locale объект локали
 */
fun forceLocaleInApp(context: Context, locale: Locale?) {
    Locale.setDefault(locale)
    context.resources?.let {
        val configuration = it.configuration
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            configuration.setLocale(locale)
            context.createConfigurationContext(configuration)
        } else {
            configuration.locale = locale
            it.updateConfiguration(configuration, it.displayMetrics)
        }
    }
}

fun copyToClipboard(context: Context, label: String?, text: String?) {
    val clipboard = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager?
            ?: throw NullPointerException(ClipboardManager::class.java.simpleName + " is null")
    val clip = ClipData.newPlainText(label, text)
    clipboard.primaryClip = clip
}

fun getDisplayMetrics(context: Context): DisplayMetrics {
    val wm = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
            ?: throw java.lang.NullPointerException(WindowManager::class.java.simpleName + "")
    val outMetrics = DisplayMetrics()
    wm.defaultDisplay.getMetrics(outMetrics)
    return outMetrics
}

/**
 * Converts pixel value to dp value
 */
fun convertPxToDp(px: Float, context: Context): Float {
    return px / context.resources.displayMetrics.density
}

@JvmOverloads
fun convertAnyToPx(
        value: Float,
        unit: Int = TypedValue.COMPLEX_UNIT_DIP,
        context: Context
): Float {
    // OR simply px = value * density (if DIP)
    return if (value <= 0) {
        0f
    } else {
        TypedValue.applyDimension(unit, value, context.resources.displayMetrics)
    }
}

fun getScreenType(context: Context): DeviceType {
    val outMetrics = getDisplayMetrics(context)
    val deviceType: DeviceType
    val shortSize = Math.min(outMetrics.heightPixels, outMetrics.widthPixels)
    val shortSizeDp = shortSize * DisplayMetrics.DENSITY_DEFAULT / outMetrics.densityDpi
    deviceType = if (shortSizeDp < 600) { // 0-599dp: "phone" UI with a separate status & navigation bar
        DeviceType.PHONE
    } else if (shortSizeDp < 720) { // 600-719dp: "phone" UI with modifications for larger screens
        DeviceType.HYBRID
    } else { // 720dp: "tablet" UI with a single combined status & navigation bar
        DeviceType.TABLET
    }
    return deviceType
}

fun isTabletUI(con: Context): Boolean {
    return getScreenType(con) == DeviceType.TABLET
}

enum class DeviceType {
    PHONE, HYBRID, TABLET
}
